### PR TITLE
 Upload a testcase's IO files

### LINKED ISSUE #272: Upload (replace) a testcase's IO. [PUT] /api/problems/{problemId}/testcases/IO
`[PUT] /api/problems/{problemId}/testcases/{testcaseId}/IO`

Body
---
Content-type: `application/zip`

1. If the `testcasesIO` exists, you must remove it before you save the new `testcaseIOs`

Response
---
200 if succeed
Body:  `text/plain`
the `testcaseIOsFileId` (from GridFs) should be included in the body

404 if the problem is not found

### PR DESCRIPTION
**Pull request by Johnny850807**
resolve #272 
 Upload a testcase's IO files

### CODE REVIEW

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java [hunk 1]
```diff
@@ -215,7 +191,64 @@ public void saveTags(List<String> tagList) {
...
+        testcase.setTestcaseIO(ioFiles);
+        Problem savedProblem = save(problem);
+        removeFileByFileId(originalIoFileId);
+        return savedProblem;
+    }
+
+    private InputStream compressTestcaseIoFiles(TestcaseIO.Files ioFiles) throws IOException {
+        PipedInputStream pipedIn = new PipedInputStream();
+        try (PipedOutputStream pipedOut = new PipedOutputStream(pipedIn);
+             ZipOutputStream zipos = new ZipOutputStream(pipedOut)) {
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/repositories/MongoProblemRepository.java:236]: NIT: PipedOutputStream、ZipOutputStream -> var

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

File: Domain/Primitives/src/main/java/tw/waterball/judgegirl/primitives/problem/Problem.java [hunk 2]
```diff
@@ -57,39 +56,27 @@
...
         setDescription(description);
-        this.languageEnvs = requireNonNull(languageEnvs);
+        this.languageEnvs = new HashMap<>(languageEnvs);
         this.outputMatchPolicyPluginTag = requireNonNull(outputMatchPolicyPluginTag);
-        this.filterPluginTags = requireNonNull(filterPluginTags);
-        setTags(tags);
-        this.testcases = requireNonNull(testcases);
+        this.filterPluginTags = new HashSet<>(filterPluginTags);
+        setTags(new ArrayList<>(tags));
+        this.testcases = new ArrayList<>(testcases);
```
[the remark this answers is not in the document]
Comment (Johnny850807) [reply Domain/Primitives/src/main/java/tw/waterball/judgegirl/primitives/problem/Problem.java:77]: Yes, I want it to throw NullPointerException, this is on purpose :)

File: Domain/Primitives/src/main/java/tw/waterball/judgegirl/primitives/problem/TestcaseIO.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import java.util.Set;
+
+import static tw.waterball.judgegirl.commons.utils.StreamUtils.mapToSet;
+import static tw.waterball.judgegirl.commons.utils.validations.ValidationUtils.validate;
+
+/**
+ * @author - johnny850807@gmail.com (Waterball)
+ */
+@Data
+@Getter
```
[the remark this answers is not in the document]
Comment (Johnny850807) [reply Domain/Primitives/src/main/java/tw/waterball/judgegirl/primitives/problem/TestcaseIO.java:21]: Nice catch

**Review by Wally5077 — APPROVED** [review]
LGTM

**Pull request merged**